### PR TITLE
feat(skore): Add support for CrossValidationReport in ConfusionMatrixDisplay

### LINKED ISSUE #2218: feat(skore): Add support for CrossValidationReport in ConfusionMatrixDisplay
### What would you like to say?

Add the std deviation across the folds in the cells of the `sns.heatmap`.
Part of #2159.

### PR DESCRIPTION
**Pull request by GaetandeCast**
Closes #2218.

Here is how the matrix looks like with a few different parameter combinations:

- default look:
image

- with `normalize="all"`:
image

- with `threshold_value=0.95`
image

- with `heatmap_kwargs={"cmap": "Greens", "fmt": ".2e"}`
image

- in multi-class classification
image

### CODE REVIEW

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 1]
```diff
@@ -288,74 +323,84 @@ def _compute_data_for_display(
         display : ConfusionMatrixDisplay
             The confusion matrix display.
         """
-        y_true_values = y_true[0].y
-        y_pred_values = y_pred[0].y
+        pos_label_validated = cls._validate_from_predictions_params(
+            y_true, y_pred, ml_task=ml_task, pos_label=pos_label
+        )
```
Comment (GaetandeCast) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py:314]: Validating the `pos_label` could have been done earlier and is not directly related to the new feature of this PR.
It is useful as it allows us to infer the `pos_label` when it is not directly set by the user, which means we always have the asterisk on the `pos_label` added in #2177.

File: skore/tests/unit/displays/confusion_matrix/test_estimator.py [hunk 2]
```diff
@@ -199,16 +200,17 @@ def test_heatmap_kwargs_override(pyplot, forest_binary_classification_with_train
 
     display = report.metrics.confusion_matrix()
     display.plot(heatmap_kwargs={"annot": False})
-    assert len(display.ax_.texts) == 0
+    # There is still the pos_label annotation
+    assert len(display.ax_.texts) == 1
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/confusion_matrix/test_estimator.py:204]: With the change to `pos_label` mentioned above, we now always have a text element in the plot: the `pos_label` annotation.

File: skore/tests/unit/displays/confusion_matrix/test_estimator.py [hunk 3]
```diff
@@ -199,16 +200,17 @@ def test_heatmap_kwargs_override(pyplot, forest_binary_classification_with_train
...
     display = report.metrics.confusion_matrix()
     display.plot(heatmap_kwargs={"fmt": "d"})
     for text in display.ax_.texts:
         assert "." not in text.get_text() or text.get_text().endswith(".0")
 
     display = report.metrics.confusion_matrix()
-    display.plot(heatmap_kwargs={"cbar": False})
-    assert len(display.figure_.axes) == 1
+    display.plot(heatmap_kwargs={"cbar": True})
+    assert len(display.figure_.axes) == 2
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/confusion_matrix/test_estimator.py:213]: We changed the default value of `"cbar"` to `False` in #2177 so it makes more sense to switch it to `True` in this test.

File: skore/tests/unit/displays/confusion_matrix/test_estimator.py [hunk 4]
```diff
@@ -259,8 +261,8 @@ def test_plot_attributes(pyplot, forest_binary_classification_with_train_test):
 
     xticklabels = [label.get_text() for label in display.ax_.get_xticklabels()]
     yticklabels = [label.get_text() for label in display.ax_.get_yticklabels()]
-    assert xticklabels == [str(label) for label in display.display_labels]
-    assert yticklabels == [str(label) for label in display.display_labels]
+    assert xticklabels == ["0", "1*"]
+    assert yticklabels == ["0", "1*"]
```
Comment (GaetandeCast) [inline skore/tests/unit/displays/confusion_matrix/test_estimator.py:265]: also due to the `pos_label` change

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 5]
```diff
@@ -144,7 +146,7 @@ def _plot_matplotlib(
         else:
             raise NotImplementedError(
                 "`ConfusionMatrixDisplay` is only implemented for "
-                "`EstimatorReport` for now."
+                "`EstimatorReport` and `CrossValidationReport` for now."
             )
 
     def _plot_single_estimator(
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: inline the method into _plot_matplotlib, or change the name (since this method is no longer specific to single estimators)

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 6]
```diff
@@ -288,74 +323,84 @@ def _compute_data_for_display(
...
+            n_cells = n_classes * n_classes
+
+            true_labels = np.tile(np.repeat(display_labels, n_classes), n_thresholds)
+            pred_labels = np.tile(np.tile(display_labels, n_classes), n_thresholds)
+            threshold_values = np.repeat(thresholds, n_cells)
+
+            counts = cms.reshape(-1)
+            normalized_true_values = cm_true.reshape(-1)
+            normalized_pred_values = cm_pred.reshape(-1)
+            normalized_all_values = cm_all.reshape(-1)
```
Comment (auguste-probabl) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: nit: Move these closer to where `cm*` are defined

File: skore/tests/unit/displays/confusion_matrix/test_cross_validation.py [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+    assert display.ax_.get_xticklabels()[1].get_text() == "B*"
+    assert display.ax_.get_yticklabels()[1].get_text() == "B*"
+
+
+def test_plot_multiclass_no_threshold_in_title(
+    pyplot, forest_multiclass_classification_data
+):
+    """Check that multiclass classification does not show threshold in title."""
+    (estimator, X, y), cv = forest_multiclass_classification_data, 3
+    report = CrossValidationReport(estimator, X=X, y=y, splitter=cv)
```
Comment (auguste-probabl) [inline skore/tests/unit/displays/confusion_matrix/test_cross_validation.py]:
```suggestion
    estimator, X, y = forest_multiclass_classification_data
    report = CrossValidationReport(estimator, X=X, y=y, splitter=3)
```

Comment (auguste-probabl) [reply skore/tests/unit/displays/confusion_matrix/test_cross_validation.py]: same for other tests

Comment (auguste-probabl) [reply skore/tests/unit/displays/confusion_matrix/test_cross_validation.py]: also maybe making `report` a fixture would make sense

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 6 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: Done. I also added the use of the `out` parameter to avoid a numpy warning.

File: skore/tests/unit/displays/confusion_matrix/test_cross_validation.py (hunk 7 above)
Comment (GaetandeCast) [reply skore/tests/unit/displays/confusion_matrix/test_cross_validation.py]: Done, I created two fixtures that create a CrossValidationReport with 3 splits using `forest_multiclass_classification_data`.

File: skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py [hunk 8]
```diff
@@ -1116,7 +1117,11 @@ def _get_display(
             cache_key = None
         else:
             cache_key_parts: list[Any] = [self._parent._hash, display_class.__name__]
-            cache_key_parts.extend(display_kwargs.values())
+            for kwarg in display_kwargs.values():
+                # NOTE: We cannot use lists in cache keys because they are not hashable
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py]: Which parameter could be a list here?

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py [hunk 9]
```diff
@@ -12,11 +12,14 @@
...
-from skore._sklearn._plot.utils import _validate_style_kwargs
+from skore._sklearn._plot.utils import (
+    _ClassifierCurveDisplayMixin,
+    _validate_style_kwargs,
+)
 from skore._sklearn.types import MLTask, PositiveLabel, ReportType, YPlotData
 
 
-class ConfusionMatrixDisplay(DisplayMixin):
+class ConfusionMatrixDisplay(_ClassifierCurveDisplayMixin, DisplayMixin):
```
Comment (glemaitre) [inline skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: It is weird to inherit from `_ClassifierCurveDisplayMixin` because a `ConfusionMatrixDisplay` is not a curve.

I assume that we want to use the `_validate_from_predictions_params` method which is useful for classification in general.

I would suggest to extract the class method into another mixin called `_ClassifierDisplayMixin` and inherit from it and make `_ClassifierCurveDisplayMixin` inherit from it as well.

File: skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py (hunk 8 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py]: `display_labels`, its computed as `self._parent.estimator_reports_[0].estimator_.classes_.tolist()`

Comment (glemaitre) [reply skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py]: I would prefer that we pass a tuple with the label everywhere.

It avoids this additional loop and since the label are not supposed to change, having an immutable is fine.

Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_cross_validation/metrics_accessor.py]: Ok I did it for the two accessors.

File: skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py (hunk 9 above)
Comment (GaetandeCast) [reply skore/src/skore/_sklearn/_plot/metrics/confusion_matrix.py]: done

**Review by glemaitre — APPROVED** [review]
LGTM

**Comment by GaetandeCast** [comment]
@auguste-probabl is it good for you too ?

**Review by auguste-probabl — APPROVED** [review]
